File: docs/perry-main-src/perry-main/benchmarks/json_polyglot/bench_field_access.cpp
// JSON parse-and-iterate polyglot benchmark — C++ (nlohmann/json).
// 10k records, ~1 MB blob, 50 iterations.
// Per iteration: parse → sum every record's nested.x → stringify.
// Identical workload to bench_field_access.ts/.go/.rs/.swift/.kt.

#include <chrono>
#include <iostream>
#include <string>
#include <vector>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

int main() {
    json items = json::array();
    for (int i = 0; i < 10000; ++i) {
        json tags = json::array();
        tags.push_back("tag_" + std::to_string(i % 10));
        tags.push_back("tag_" + std::to_string(i % 5));
        items.push_back({
            {"id", i},
            {"name", "item_" + std::to_string(i)},
            {"value", static_cast<double>(i) * 3.14159},
            {"tags", tags},
            {"nested", { {"x", i}, {"y", i * 2} }}
        });
    }
    std::string blob = items.dump();

    // Warmup
    for (int i = 0; i < 3; ++i) {
        json parsed = json::parse(blob);
        long long warm_sum = 0;
        for (const auto& item : parsed) {
            warm_sum += item["nested"]["x"].get<long long>();
        }
        (void)parsed.dump();
        (void)warm_sum;
    }

    constexpr int iterations = 50;
    auto start = std::chrono::steady_clock::now();

    long long checksum = 0;
    for (int iter = 0; iter < iterations; ++iter) {
        json parsed = json::parse(blob);
        long long sum = 0;
        for (const auto& item : parsed) {
            sum += item["nested"]["x"].get<long long>();
        }
        checksum += sum;
        std::string reStringified = parsed.dump();
        checksum += static_cast<long long>(reStringified.size());
    }

    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::steady_clock::now() - start).count();
    std::cout << "ms:" << elapsed << "\n";
    std::cout << "checksum:" << checksum << "\n";
    return 0;
}

File: docs/perry-main-src/perry-main/benchmarks/json_polyglot/bench_field_access.rs
// JSON parse-and-iterate polyglot benchmark — Rust (serde_json).
// 10k records, ~1 MB blob, 50 iterations.
// Per iteration: parse → sum every record's nested.x → stringify.
// Identical workload to bench_field_access.ts/.go/.swift/.cpp/.kt.

use serde::{Deserialize, Serialize};
use std::time::Instant;

#[derive(Serialize, Deserialize)]
struct Nested {
    x: i64,
    y: i64,
}

#[derive(Serialize, Deserialize)]
struct Item {
    id: i64,
    name: String,
    value: f64,
    tags: Vec<String>,
    nested: Nested,
}

fn main() {
    let mut items: Vec<Item> = Vec::with_capacity(10_000);
    for i in 0..10_000i64 {
        items.push(Item {
            id: i,
            name: format!("item_{}", i),
            value: (i as f64) * 3.14159,
            tags: vec![format!("tag_{}", i % 10), format!("tag_{}", i % 5)],
            nested: Nested { x: i, y: i * 2 },
        });
    }
    let blob = serde_json::to_string(&items).expect("stringify blob");

    // Warmup
    for _ in 0..3 {
        let parsed: Vec<Item> = serde_json::from_str(&blob).expect("parse");
        let mut warm_sum: i64 = 0;
        for item in &parsed {
            warm_sum += item.nested.x;
        }
        let _ = serde_json::to_string(&parsed).expect("stringify");
        let _ = warm_sum;
    }

    const ITERATIONS: usize = 50;
    let start = Instant::now();

    let mut checksum: i64 = 0;
    for _ in 0..ITERATIONS {
        let parsed: Vec<Item> = serde_json::from_str(&blob).expect("parse");
        let mut sum: i64 = 0;
        for item in &parsed {
            sum += item.nested.x;
        }
        checksum += sum;
        let re_stringified = serde_json::to_string(&parsed).expect("stringify");
        checksum += re_stringified.len() as i64;
    }

    let elapsed = start.elapsed().as_millis();
    println!("ms:{}", elapsed);
    println!("checksum:{}", checksum);
}
